frontend/src/models/propertyDefinitionsModel.ts

```diff
@@ -1,163 +1,223 @@
     },
-]
-
-const normaliseToArray = (
-    valueToFormat: Exclude<PropertyFilterValue, null>
-): {
-    valueWasReceivedAsArray: boolean
-    arrayOfPropertyValues: (string | number)[]
-} => {
-    if (Array.isArray(valueToFormat)) {
-        return { arrayOfPropertyValues: valueToFormat, valueWasReceivedAsArray: true }
-    } else {
-        return { arrayOfPropertyValues: [valueToFormat], valueWasReceivedAsArray: false }
-    }
 }
 
 export type FormatPropertyValueForDisplayFunction = (
     propertyName?: BreakdownKeyType,
     valueToFormat?: PropertyFilterValue
 ) => string | string[] | null
 
-export const propertyDefinitionsModel = kea<propertyDefinitionsModelType>({
-    path: ['models', 'propertyDefinitionsModel'],
-    actions: () => ({
-        loadPropertyDefinitions: (initial = false) => ({ initial }),
-        updatePropertyDefinition: (property: PropertyDefinition) => ({ property }),
-    }),
-    loaders: ({ values }) => ({
-        propertyStorage: [
-            { results: [], next: null, count: 0 } as PropertyDefinitionStorage,
-            {
-                loadPropertyDefinitions: async ({ initial }, breakpoint) => {
-                    const url = initial
-                        ? 'api/projects/@current/property_definitions/?limit=5000'
-                        : values.propertyStorage.next
-                    if (!url) {
-                        throw new Error('Incorrect call to propertyDefinitionsLogic.loadPropertyDefinitions')
-                    }
-                    const propertyStorage = await api.get(url)
-                    breakpoint()
-                    return {
-                        count: propertyStorage.count,
-                        results: [...values.propertyStorage.results, ...propertyStorage.results],
-                        next: propertyStorage.next,
-                    }
-                },
-            },
-        ],
+/** Update cached property definition metadata */
+export const updatePropertyDefinitions = (
+    propertyDefinitions: PropertyDefinition[] | PropertyDefinitionStorage
+): void => {
+    propertyDefinitionsModel.findMounted()?.actions.updatePropertyDefinitions(propertyDefinitions)
+}
+
+/** Update cached property definition metadata */
+export const updatePropertyDefinition = (propertyDefinitions: PropertyDefinition): void => {
+    propertyDefinitionsModel.findMounted()?.actions.updatePropertyDefinition(propertyDefinitions)
+}
+
+/** Schedules a background task with a 10ms debounce to fetch property definitions */
+const checkOrLoadPropertyDefinition = (
+    propertyName: BreakdownKeyType | undefined,
+    propertyDefinitionStorage: PropertyDefinitionStorage
+): void => {
+    // first time we see this, schedule a fetch
+    if (typeof propertyName === 'string' && !(propertyName in propertyDefinitionStorage)) {
+        window.setTimeout(
+            () => propertyDefinitionsModel.findMounted()?.actions.loadPropertyDefinitions([propertyName]),
+            0
```

Comment:
Comment says 10ms code says 0. A mistake or arcane JS :)?

Reply:
Updated the comment. This line still gets a 0ms timeout to not e.g. execute when React is rendering. Once the action is called and there's something to fetch, we add a 10ms debounce.